### ISSUE TITLE
How to fail on insert if key is already present

### ISSUE THREAD

**Issue by cameron1024** [xacrimon/dashmap#196]
## TLDR

Is there a way to make `.insert()` fail if the key is already present?

Hi, I'm trying to generate random keys for a `DashMap` but my current strategy has a data race of sorts. It's not memory-unsafe (no UB or anything), but it breaks an invariant of the struct I'm trying to build:
```rust
struct MyStruct {
  map: DashMap<u64, String>,
}

impl MyStruct {
  fn new_with_random_key(&self, data: String) {
    for _ in 0..10 {
      let key = rand::random();
      if !self.map.contains_key(&key) {
        //  another thread may do an insert for the same key here
        self.map.insert(key, data);
        return;
      }
    }
    panic!("too many attempts");
  }
}
```
Currently I'm getting around this by just using `u64`s as keys, and hoping I don't get any collisions, but this makes me a bit uneasy. If another thread happens to generate the same key, and the insert happens at the point where the comment is, then that thread will "succeed", but its data will be immediately dropped when this thread finishes its insert, which breaks an invariant I'm trying to maintain (that map entries are only removed when an explicit `MyStruct::remove` function is called).

Essentially I'm looking for some way to make the `contains_key` check and the `insert` happen in a single atomic step. I've had a look through the API docs but I can't find anything that seems to do this, but maybe I'm not reading it properly.

Of course, using a `Mutex`/`RwLock` makes this trivial, but I'd really like to avoid resorting to that.  Or maybe this is an XY problem, very open to any suggestions on better/more idiomatic ways to achieve the same semantics.

Thanks :grin:

**Comment by jbabyhacker** [comment]
I concur, there is `remove_if`, but I think this library would benefit from an `insert_if`. I am running into a similar situation.

**Comment by apendleton** [comment]
@cameron1024 I think you can use the [`Entry` API](https://docs.rs/dashmap/latest/dashmap/struct.DashMap.html#method.entry). The `Entry` that you get back will either be `Vacant` or `Occupied`, and you can call `insert` on the `VacantEntry` variant if that's what you get back, or generate a new key otherwise. While you hold the `Entry`, it shouldn't be possible for other threads to write to that key, if I understand the underlying guarantees correctly (I think you'll hold a lock on that shard).

**Comment by xacrimon** [comment]
Yup. This is doable with the Entry API.
